routing/index_router.cpp

```diff
@@ -154,37 +154,60 @@ void PushPassedSubroutes(Checkpoints const & checkpoints, vector<Route::Subroute
 namespace routing
 {
 // IndexRouter::BestEdgeComparator ----------------------------------------------------------------
-IndexRouter::BestEdgeComparator::BestEdgeComparator(m2::PointD const & point, m2::PointD const & direction)
-  : m_point(point), m_direction(direction)
+IndexRouter::BestEdgeComparator::BestEdgeComparator(m2::PointD const & point,
+                                                    m2::PointD const & direction,
+                                                    IsOnewayFn const & isOnewayFn)
+  : m_point(point), m_direction(direction), m_isOnewayFn(isOnewayFn)
 {
 }
 
 bool IndexRouter::BestEdgeComparator::Compare(Edge const & edge1, Edge const & edge2) const
 {
-  if (!m_direction.IsAlmostZero())
+  if (IsDirectionValid())
   {
-    bool const isEdge1Parallel = IsAlmostParallel(edge1);
-    if (isEdge1Parallel != IsAlmostParallel(edge2))
+    bool const isEdge1Parallel = IsAlmostConformed(edge1);
+    if (isEdge1Parallel != IsAlmostConformed(edge2))
       return isEdge1Parallel;
   }
 
   return GetSquaredDist(edge1) < GetSquaredDist(edge2);
 }
 
-bool IndexRouter::BestEdgeComparator::IsAlmostParallel(Edge const & edge) const
+bool IndexRouter::BestEdgeComparator::IsAlmostConformed(Edge const & edge) const
 {
-  CHECK(!m_direction.IsAlmostZero(), ());
+  bool forward;
+  if (!IsAlmostCollinear(edge, forward))
+    return false;
+
+  // Note. In case of oneway edges |edge| is conformed to |m_direction| iff vector of |edge|
+  // and vector |m_direction| are almost collinear and have the same direction.
+  return m_isOnewayFn(edge.GetFeatureId()) ? forward : true;
+}
 
+bool IndexRouter::BestEdgeComparator::IsAlmostCollinear(Edge const & edge, bool & forward) const
```

Comment:
|BestEdgeComparator::m_direction| vector and vector which is made by |edge| param.